src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DNodeViewModel.cs

```diff
@@ -0,0 +1,89 @@
+﻿using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Windows;
+using Autodesk.DesignScript.Interfaces;
+using Dynamo.Models;
+using Dynamo.Utilities;
+
+namespace Dynamo.Wpf.ViewModels.Watch3D
+{
+    public class HelixWatch3DNodeViewModel : HelixWatch3DViewModel
+    {
+        private NodeModel node;
+
+        public static HelixWatch3DNodeViewModel Start(NodeModel node, Watch3DViewModelStartupParams parameters)
+        {
+            var vm = new HelixWatch3DNodeViewModel(node, parameters);
+            vm.OnStartup();
+            return vm;
+        }
+
+        private HelixWatch3DNodeViewModel(NodeModel node, Watch3DViewModelStartupParams parameters):
+            base(parameters)
+        {
+            this.node = node;
+            IsResizable = true;
+
+            RegisterPortEventHandlers(node);
+        }
+
+        protected override void PortConnectedHandler(PortModel arg1, ConnectorModel arg2)
+        {
+            UpdateUpstream();
+        }
+
+        private void UpdateUpstream()
+        {
+            OnClear();
+
+            var gathered = new List<NodeModel>();
+            node.VisibleUpstreamNodes(gathered);
+            if (gathered.Any())
+            {
+                gathered.ForEach(n => n.IsUpdated = true);
+            }
+
+            gathered.ForEach(n => n.RequestVisualUpdateAsync(model.Scheduler, model.EngineController, factory));
+        }
+
+        protected override void PortDisconnectedHandler(PortModel obj)
+        {
+            OnClear();
+        }
+
+        protected override void OnNodePropertyChanged(object sender, PropertyChangedEventArgs e)
+        {
+            var updatedNode = sender as NodeModel;
+
+            // Don't bother with node property changes 
+            // that are not in this branch.
+
+            if (!updatedNode.IsUpstreamOf(node))
+                return;
+
+            switch (e.PropertyName)
+            {
+                case "IsUpstreamVisible":
+                    UpdateUpstream();
+                    break;
+            }
+
+            base.OnNodePropertyChanged(sender, e);
+        }
+
+        protected override void OnUpdatedRenderPackagesAvailable(NodeModel updatedNode,
+            IEnumerable<IRenderPackage> renderPackages)
+        {
+            var visibleUpstream = new List<NodeModel>();
+            node.VisibleUpstreamNodes(visibleUpstream);
+
+            if (node == null || !visibleUpstream.Contains(updatedNode))
```

Comment:
We're checking for `node == null` here but we'd have a `NullReferenceException` on L79.